tests/test_build_data.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import os
+from parlai.core import build_data
+import unittest
+import parlai.core.testing_utils as testing_utils
+from parlai.core.params import ParlaiParser
+
+
+class TestBuildData(unittest.TestCase):
+    """Basic tests on the build_data.py download_multiprocess."""
+
+    dest_filenames = ['mnist0.tar.gz', 'mnist1.tar.gz', 'mnist2.tar.gz']
+
+    def setUp(self):
+        self.datapath = ParlaiParser().parse_args(print_args=False)['datapath']
+        for d in self.dest_filenames:
+            # Removing files if they are already there b/c otherwise it won't try to download them again
+            try:
+                os.remove(os.path.join(self.datapath, d))
+            except Exception:
+                pass
+
+    def test_download_multiprocess(self):
+        with testing_utils.capture_output() as stdout:
+            urls = [
+                'http://parl.ai/downloads/mnist/mnist.tar.gz',
+                'http://parl.ai/downloads/mnist/mnist.tar.gz.BAD',
+                'http://parl.ai/downloads/mnist/mnist.tar.gz.BAD',
+            ]
+
+            download_results = build_data.download_multiprocess(
+                urls, self.datapath, dest_filenames=self.dest_filenames
+            )
+
+        str_output = stdout.getvalue()
+        print('stdout output: %s' % str_output)
+        output_filenames = [
+            download_results[0][0],
+            download_results[1][0],
+            download_results[2][0],
+        ]
+        output_statuses = [
+            download_results[0][1],
+            download_results[1][1],
+            download_results[2][1],
+        ]
+        self.assertEquals(
+            output_filenames,
+            ['mnist0.tar.gz', 'mnist1.tar.gz', 'mnist2.tar.gz'],
```

Comment:
so we still create the files even if the downloads fail?

Reply:
Function returns tuples of (filename, status, error msg). File will not be created if status != 200 but we need to return same N of tuples with statuses, etc as the input. Added documentation to make this clear.